### PR TITLE
fix StateUpdater memory leak issue

### PR DESCRIPTION
**Pull request by brycezhongqing**
## Summary
Recently, the `d2-proxy` experienced an `OOM` issue([more info can find here](https://drive.google.com/file/d/17tKFHzd3MX_bUBByA3nTNl_rnYUKt5Bo/view?usp=sharing)). Investigation revealed several reasons for this, with one stemming from rest.li **_StateUpdater_**. The **_StateUpdater_** utilizes **__executorService_** to run tasks periodically, **but these tasks persist despite no longer being necessary**. The absence of a proper **_shutdown_** method prevents task cancellation, resulting in numerous unused StateUpdater objects lingering and causing memory leaks.

## Test done
We used `toki` to test.
- Step1:   release the rest.li snapshot locally. `./scripts/local-release -s`

- Step2:  bump up rest.li snapshot in toki locally
- Step3:  `mint build && mint build-cfg -f qei-ltx1 && mint deploy -w toki-war -f qei-ltx1`

- Step4: `cd /export/content/lid/apps/toki/i001`  and `bin/control stop` to stop service

Then u will find the following `shutdown log` (Please note, during debug, we could change the log level from `debug` to `info`)

```
2024/03/19 15:32:37.721 INFO [StateUpdater] [__r2d2DefaultClient__.r2d2Client.d2Executor-4-1] [toki-war] [AAYUCw0g7l/m8yJa5/+xCg==] Shutting down the state updater for service: dataVaultAccessControlList
2024/03/19 15:32:37.721 INFO [StateUpdater] [__r2d2DefaultClient__.r2d2Client.d2Executor-4-1] [toki-war] [AAYUCw0g7l/m8yJa5/+xCg==] Shutting down the state updater for service: dataVaultAclChanges
2024/03/19 15:32:37.721 INFO [StateUpdater] [__r2d2DefaultClient__.r2d2Client.d2Executor-4-1] [toki-war] [AAYUCw0g7l/m8yJa5/+xCg==] Shutting down the state updater for service: lixClientContexts
2024/03/19 15:32:37.721 INFO [StateUpdater] [__r2d2DefaultClient__.r2d2Client.d2Executor-4-1] [toki-war] [AAYUCw0g7l/m8yJa5/+xCg==] Shutting down the state updater for service: sloThresholds
2024/03/19 15:32:37.721 INFO [StateUpdater] [__r2d2DefaultClient__.r2d2Client.d2Executor-4-1] [toki-war] [AAYUCw0g7l/m8yJa5/+xCg==] Shutting down the state updater for service: tokiBackend
```

 

[toki-war.log](https://github.com/linkedin/rest.li/files/14657932/toki-war.log)

## Follow up
bump up the `rest.li` version for `d2-proxy`  later and check memory usage.
This is very intersting bug investigation, will write a `one-pager` for this :)

### CODE REVIEW

**Review by bohhyang — COMMENTED** [review]
great finding! thanks! will wait for the d2-proxy test result to ship.

**Comment by brycezhongqing** [comment]
> great finding! thanks! will wait for the d2-proxy test result to ship.

Hi, @bohhyang As the d2-proxy in ei not use too much memory. So my plan is to merge this rest.li PR and check the performance improvement of d2-proxy in prod(Which is using over 6GB memory now). What do u think ?

**Comment by bohhyang** [comment]
> > great finding! thanks! will wait for the d2-proxy test result to ship.
> 
> Hi, @bohhyang As the d2-proxy in ei not use too much memory. So my plan is to merge this rest.li PR and check the performance improvement of d2-proxy in prod(Which is using over 6GB memory now). What do u think ?

sounds good.

**Review by bohhyang — APPROVED** [review]

File: d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/StateUpdater.java [hunk 1]
```diff
@@ -28,10 +28,7 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ConcurrentMap;
-import java.util.concurrent.ScheduledExecutorService;
-import java.util.concurrent.TimeUnit;
+import java.util.concurrent.*;
```
Comment (bohhyang) [inline d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/StateUpdater.java:31]: this should expand each imports. (it's a code style setting in IntelliJ preferences you may need to fix)

Comment (brycezhongqing) [reply d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/StateUpdater.java]: Oh. Got it. Updated. Thanks

**Pull request merged**